apps/prairielearn/src/tests/render.test.ts

```diff
@@ -0,0 +1,454 @@
+        } as unknown as Submission;
+
+        const { courseIssues: parseIssues, data: parsedData } = await questionModule.parse(
+          submissionInput,
+          variant as unknown as Variant, // Cast needed
+          question,
+          course,
+        );
+        assert.isEmpty(
+          parseIssues,
+          `Parse courseIssues should be empty for input ${JSON.stringify(
+            submissionInput.submitted_answer,
+            undefined,
+            2,
+          )}, but it was ${JSON.stringify(parseIssues)}`,
+        );
+        // console.log(parsedData);
+        // if ((parsedData.format_errors as any).length > 0) {
+        //   console.log('Format errors for question directory:', relativePath);
+        // parsedData.format_errors = {};
+        // parsedData.submitted_answer = structuredClone(variant.true_answer);
+        // parsedData.gradable = true;
+        // }
+        // console.log(parsedData);
+        for (const issue of Object.keys(parseIssues)) {
+          if (issue === '_files') {
+            console.log("Can't handle file questions yet");
+            return;
+          }
+        }
+        for (const issue of Object.keys(parsedData.format_errors ?? {})) {
+          if (issue === '_files') {
+            console.log("Can't handle file questions yet");
+            return;
+          }
+        }
+
+        assert.isEmpty(
+          parsedData.format_errors ?? {},
+          `Parse format_errors should be empty
+          but it was ${JSON.stringify(parsedData.format_errors, undefined, 2)} for submission ${JSON.stringify(
+            submissionInput.submitted_answer,
+            undefined,
+            2,
+          )}`,
+        );
+
+        // Update submission with parsed data
+        const submissionForGrading = { ...parsedData };
+        // if (Object.keys(submissionForGrading.format_errors ?? {}).length > 0) {
+        //   submissionForGrading.gradable = false;
+        // }
+
+        // assert.isTrue(submissionForGrading.gradable, 'Submission should be gradable after parse');
+
+        // 5. Grade
+        const { courseIssues: gradeIssues, data: gradeData } = await questionModule.grade(
+          submissionForGrading as unknown as Submission, // Cast needed
+          variant as unknown as Variant, // Cast needed
+          question,
+          course,
+        );
+
+        for (const issue of gradeIssues) {
+          if (issue.data.outputStderr) {
+            assert.fail(
+              JSON.stringify(submissionForGrading.submitted_answer, undefined, 2) +
+                '\n' +
+                issue.data.outputStderr,
+            );
+          }
+        }
+        assert.isEmpty(
+          gradeIssues,
+          `Grade courseIssues should be empty but got ${JSON.stringify(gradeIssues)}`,
+        );
+        assert.isEmpty(gradeData.format_errors ?? {}, 'Grade format_errors should be empty');
+        if ((gradeData.true_answer as any).length > 0) {
```

Comment:
Some questions don't even have a correct answer, so it's not worth checking if the input was marked as correct.